cmd/admission-webhook/main.go

```diff
@@ -0,0 +1,269 @@
+		"If omitted, the default Go cipher suites will be used."+
+		"Note that TLS 1.3 ciphersuites are not configurable.")
+
+	flagset.StringVar(&cfg.LogLevel, "log-level", logging.LevelInfo, fmt.Sprintf("Log level to use. Possible values: %s", strings.Join(logging.AvailableLogLevels, ", ")))
+	flagset.StringVar(&cfg.LogFormat, "log-format", logging.FormatLogFmt, fmt.Sprintf("Log format to use. Possible values: %s", strings.Join(logging.AvailableLogFormats, ", ")))
+
+	_ = flagset.Parse(os.Args[1:])
+
+	logger, err := logging.NewLogger(cfg.LogLevel, cfg.LogFormat)
+	if err != nil {
+		stdlog.Fatal(err)
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	wg, ctx := errgroup.WithContext(ctx)
+
+	listener, err := net.Listen("tcp", cfg.ListenAddress)
+	if err != nil {
+		level.Error(logger).Log("msg", "failed to start required HTTP listener", "err", err)
+		os.Exit(1)
+	}
+
+	tlsConf, err := loadTLSConfigFromFlags(ctx, logger, wg)
+	if err != nil {
+		level.Error(logger).Log("msg", "failed to build TLS config", "err", err)
+		os.Exit(1)
+	}
+
+	server := newSrv(logger, tlsConf)
+	wg.Go(func() error {
+		return server.run(listener)
+	})
+
+	term := make(chan os.Signal, 1)
+	signal.Notify(term, os.Interrupt, syscall.SIGTERM)
+
+	select {
+	case sig := <-term:
+		level.Info(logger).Log("msg", "Received signal, exiting gracefully...", "signal", sig.String())
+	case <-ctx.Done():
+	}
+
+	if err := server.shutdown(ctx); err != nil {
+		level.Warn(logger).Log("msg", "Server shutdown error", "err", err)
+	}
+
+	cancel()
+	if err := wg.Wait(); err != nil {
+		level.Warn(logger).Log("msg", "Unhandled error received. Exiting...", "err", err)
+		os.Exit(1)
+	}
+}
+
+func (s *srv) run(listener net.Listener) error {
+	log := log.With(s.logger, "address", listener.Addr().String())
+
+	if s.s.TLSConfig != nil {
+		level.Info(log).Log("msg", "Starting TLS enabled server")
+		if err := s.s.ServeTLS(listener, "", ""); err != http.ErrServerClosed {
+			return err
+		}
+		return nil
+	}
+
+	level.Info(log).Log("msg", "Starting insecure server")
+	if err := s.s.Serve(listener); err != http.ErrServerClosed {
+		return err
+	}
+	return nil
+}
+
+func (s *srv) shutdown(ctx context.Context) error {
+	return s.s.Shutdown(ctx)
+}
+
+func newSrv(logger log.Logger, tlsConf *tls.Config) *srv {
+	mux := http.NewServeMux()
+	admit := admission.New(log.With(logger, "component", "admissionwebhook"))
+	admit.Register(mux)
+
+	ruleValidationTriggeredCounter := prometheus.NewCounter(prometheus.CounterOpts{
+		Name: "prometheus_operator_rule_validation_triggered_total",
+		Help: "Number of times a PrometheusRule object triggered validation",
+	})
+
+	alertManagerConfigValidationTriggered := prometheus.NewCounter(prometheus.CounterOpts{
+		Name: "prometheus_operator_alertmanager_config_validation_triggered_total",
+		Help: "Number of times an AlertManagerConfig object triggered validation",
+	})
```

Comment:
thinking more about it, I would be ok dropping them for now. I could see metrics on the HTTP handlers in the future though (e.g. request counter and latency per status code and path).

Reply:
yeah I am totally in favour of that. I'll wait to see if anyone other maintainer has strong opinions on this one before deprecating these from existing webhook and dropping them here entirely. If no objections then I think we are in agreement with the path forward.

Reply:
Since we don't have any alerts on this, I am fine with dropping this and and including metrics on HTTP handlers in the future :+1: